app/models/post_redirect.rb

```diff
@@ -91,6 +116,17 @@ def generate_token
   # There is a separate token to use in the URL if we send a confirmation
   # email.
   def generate_email_token
-    self.email_token = PostRedirect.generate_random_token unless email_token
+    case circumstance
+    when 'change_email'
+      self.email_token ||= PostRedirect.generate_verifiable_token(
+        user: user, value: user.email
+      )
+    when 'change_password'
+      self.email_token ||= PostRedirect.generate_verifiable_token(
+        user: user, value: user.hashed_password
+      )
+    else
+      self.email_token ||= PostRedirect.generate_random_token
```

Comment:
Could this be generated from an _unchangeable_ attribute – `User#id` say – so that we do away with the verifiable/random conditionals?

```ruby
CIRCUMSTANCE_ATTRIBUTES = {
  change_email: :email,
  change_password: :hashed_password,
  normal: :id
}

def generate_email_token
  self.email_token ||= PostRedirect.generate_verifiable_token(
    user: user, value: user.send(CIRCUMSTANCE_ATTRIBUTES[circumstance])
  )
end
```

If something like that works then I imagine `email_token_valid?` could be simplified in a similar way.

Reply:
I've kinda implemented this... if we use `User#id` then the email token will always be the same. This could meant multiple `PostRedirect` instances with the same email token. I think we would need to expire tokens some other way when they've been used.

I've kept the fallback to the old generator for now.